### PR TITLE
[Dev Portal] Revamp userstore views

### PR DESCRIPTION
**Pull request by thivi**
- The modified userstore list view
image

- The improved userstore add wizard.
image

- The revamped userstore edit view.
image

- Display SQL queries inside a code editor.
image

- Hide optional properties by default.
image

### CODE REVIEW

File: apps/developer-portal/src/constants/user-store-constants.ts [hunk 1]
```diff
@@ -16,11 +16,37 @@
...
+/**
+ * The ID of the userstore type JDBC.
+ *
+ * @constant
+ *
+ * @type {string}
+ */
 export const JDBC = "JDBC";
+
+export const UserstoreTypeDescriptions = {
```
Comment (jeradrutnam) [inline apps/developer-portal/src/constants/user-store-constants.ts:28]: `USERSTORE_TYPE_DESCRIPTIONS` do we have a plan to change all the constants uppercase ?

Comment (jeradrutnam) [reply apps/developer-portal/src/constants/user-store-constants.ts:28]: As well as we need to localize these right?

File: apps/developer-portal/src/utils/userstores.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
+import { UserStoreProperty } from "./../models/user-stores";
```
Comment (jeradrutnam) [inline apps/developer-portal/src/utils/userstores.ts]: `import` should be after the license header

File: modules/react-components/src/transfer/tranfer.tsx [hunk 3]
```diff
@@ -64,11 +64,11 @@ export const TransferComponent: FunctionComponent<PropsWithChildren<TransferComp
                                             list.props.listType === "unselected" && (
                                             <Grid.Column width={ 7 }>
                                                 <Segment
-                                                    data-testid={ `${ props[ `data-testid` ] }_unselected_groups` }
+                                                    data-testid={ `${ props[ "data-testid" ] }_unselected_groups` }
                                                     className="transfer-segment"
                                                 >
                                                     <TransferListSearch
-                                                        searchFieldTestId={ `${ props[ `data-testid` ] }_unselected_groups_search_input` }
+                                                        searchFieldTestId={ `${ props[ "data-testid" ] }_unselected_groups_search_input` }
```
Comment (jeradrutnam) [inline modules/react-components/src/transfer/tranfer.tsx:71]: Fix character length

File: modules/react-components/src/transfer/tranfer.tsx [hunk 4]
```diff
@@ -88,7 +88,7 @@ export const TransferComponent: FunctionComponent<PropsWithChildren<TransferComp
                                                 >
                                                     <Grid.Row>
                                                         <Button
-                                                            data-testid={ `${  props[ `data-testid` ] }_unselected_groups_add_button` }
+                                                            data-testid={ `${  props[ "data-testid" ] }_unselected_groups_add_button` }
```
Comment (jeradrutnam) [inline modules/react-components/src/transfer/tranfer.tsx:91]: Fix character length

File: modules/react-components/src/transfer/tranfer.tsx [hunk 5]
```diff
@@ -99,7 +99,7 @@ export const TransferComponent: FunctionComponent<PropsWithChildren<TransferComp
                                                     </Grid.Row>
                                                     <Grid.Row>
                                                         <Button
-                                                            data-testid={ `${  props[ `data-testid` ] }_unselected_groups_remove_button` }
+                                                            data-testid={ `${  props[ "data-testid" ] }_unselected_groups_remove_button` }
```
Comment (jeradrutnam) [inline modules/react-components/src/transfer/tranfer.tsx:102]: Fix character length

File: modules/react-components/src/transfer/tranfer.tsx [hunk 6]
```diff
@@ -115,11 +115,11 @@ export const TransferComponent: FunctionComponent<PropsWithChildren<TransferComp
                                             list.props.listType === "selected" && (
                                                 <Grid.Column width={ 7 } className="transfer-list-assigned-column">
                                                     <Segment
-                                                        data-testid={ `${ props[ `data-testid` ] }_selected_groups` }
+                                                        data-testid={ `${ props[ "data-testid" ] }_selected_groups` }
                                                         className="transfer-segment"
                                                     >
                                                         <TransferListSearch
-                                                            searchFieldTestId={ `${ props[ `data-testid` ] }_selected_groups_search_input` }
+                                                            searchFieldTestId={ `${ props[ "data-testid" ] }_selected_groups_search_input` }
```
Comment (jeradrutnam) [inline modules/react-components/src/transfer/tranfer.tsx:122]: Fix character length

File: apps/developer-portal/src/components/user-stores/edit/edit-user-details-userstore.tsx [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    /**
+     * The type meta data
+     */
+    type: UserstoreType;
+    /**
+     * The connection properties.
+     */
+    properties: RequiredBinary;
+}
+export const EditUserDetails = (
```
Comment (jeradrutnam) [inline apps/developer-portal/src/components/user-stores/edit/edit-user-details-userstore.tsx:50]:
```suggestion

export const EditUserDetails = (
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 8]
```diff
@@ -18,15 +18,23 @@
...
+import { USER_STORES_PATH, UserstoreTypeDisplayNames } from "../../constants";
+import { history } from "../../helpers";
+import {
+    AlertLevels,
+    CategorizedProperties,
+    UserStorePostData,
+    UserStoreProperty,
+    UserstoreType
+} from "../../models";
 import { addAlert } from "../../store/actions";
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:36]:
```suggestion
import { addAlert } from "@wso2is/core/store";
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 9]
```diff
@@ -50,19 +61,33 @@ interface AddUserStoreProps {
...
 
-    const { open, onClose } = props;
+    const { open, onClose, type } = props;
 
-    const [currentWizardStep, setCurrentWizardStep] = useState(0);
-    const [basicDetailsData, setBasicDetailsData] = useState<Map<string, FormValue>>(null);
-    const [connectionDetailsData, setConnectionDetailsData] = useState<Map<string, FormValue>>(null);
-    const [type, setType] = useState<Type>(null);
-    const [userStore, setUserStore] = useState<UserStorePostData>(null);
+    const [ currentWizardStep, setCurrentWizardStep ] = useState(0);
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:66]:
```suggestion
    const [ currentWizardStep, setCurrentWizardStep ] = useState<number>(0);
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 10]
```diff
@@ -80,48 +105,75 @@ export const AddUserStore = (props: AddUserStoreProps): ReactElement => {
...
 
     /**
-     * This saves the Basic Details values
-     * @param {Map<string, FormValue>} values Basic Details Values to be submitted
+     * This saves the General Details values along with the Type
+     * @param {Map<string, FormValue>} values Connection Details Values
      */
-    const onSubmitBasicDetails = (values: Map<string, FormValue>) => {
-        setBasicDetailsData(values);
+    const onSubmitGeneralDetails = (values: Map<string, FormValue>) => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:122]:
```suggestion
    const onSubmitGeneralDetails = (values: Map<string, FormValue>): void => {
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 11]
```diff
@@ -80,48 +105,75 @@ export const AddUserStore = (props: AddUserStoreProps): ReactElement => {
...
 
-    /**
-     * This saves the Connection Details values along with the te Type
-     * @param {Map<string, FormValue>} values Connection Details Values
-     * @param {Type} type 
-     */
-    const onSubmitConnectionDetails = (values: Map<string, FormValue>, type: Type) => {
-        setConnectionDetailsData(values);
-        setType(type);
+    const onSubmitUserDetails = (values: Map<string, FormValue>) => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:127]:
```suggestion
    const onSubmitUserDetails = (values: Map<string, FormValue>): void => {
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 12]
```diff
@@ -80,48 +105,75 @@ export const AddUserStore = (props: AddUserStoreProps): ReactElement => {
...
+        setUserDetailsData(values);
+        setCurrentWizardStep(2);
+    }
+
+    const onSubmitGroupDetails = (values: Map<string, FormValue>) => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:132]:
```suggestion
    const onSubmitGroupDetails = (values: Map<string, FormValue>): void => {
```

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 13]
```diff
@@ -80,48 +105,75 @@ export const AddUserStore = (props: AddUserStoreProps): ReactElement => {
...
+        setGroupDetailsData(values);
+    }
 
-        const data = new Map([...Array.from(basicDetailsData ?? []), ...Array.from(values ?? [])]);
+    const serializeData = () => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:136]:
```suggestion
    const serializeData = (): void => {
```

File: apps/developer-portal/src/components/user-stores/edit/edit-basic-details-user-store.tsx [hunk 14]
```diff
@@ -52,79 +58,419 @@ export const EditBasicDetailsUserStore = (
...
-                            name="description"
-                            type="textarea"
-                            required={ false }
-                            requiredErrorMessage=""
-                            placeholder="Enter a description"
-                            value={ userStore?.description }
+        <>
+            { confirmDelete && deleteConfirmation() }
+            <Forms
+                onSubmit={ (values: Map<string, FormValue>) => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/edit/edit-basic-details-user-store.tsx:148]: WDYT about separating this out to a handler function?
Check other places as well.

File: apps/developer-portal/src/components/user-stores/add-user-store.tsx [hunk 15]
```diff
@@ -50,19 +61,33 @@ interface AddUserStoreProps {
  */
 export const AddUserStore = (props: AddUserStoreProps): ReactElement => {
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/add-user-store.tsx:62]: Missing type. Check other places as well.

```suggestion
export const AddUserStore: FunctionComponent<AddUserStoreProps> = (props: AddUserStoreProps): ReactElement => {
```

File: apps/developer-portal/src/components/user-stores/wizards/general-details-userstore.tsx [hunk 16]
```diff
@@ -16,70 +16,59 @@
...
+    basicProperties: TypeProperty[];
 }
 
 /**
- * This component renders the Connection Details step of the wizard
- * @param {ConnectionDetailsPropsInterface} props
- * @return {Promise<any>}
+ * This component renders the General Details step of the wizard
+ * @param {GeneralDetailsUserstorePropsInterface} props
+ * @returns {Promise<any>}
```
Comment (brionmario) [inline apps/developer-portal/src/components/user-stores/wizards/general-details-userstore.tsx:59]: Add component type as well.
```suggestion
 * @returns {React. ReactElement}
```

File: apps/developer-portal/src/configs/routes.ts [hunk 17]
```diff
@@ -288,6 +280,24 @@ const DASHBOARD_LAYOUT_ROUTES: RouteInterface[] = [
         protected: true,
         showOnSidePanel: true
     },
+    {
```
Comment (brionmario) [inline apps/developer-portal/src/configs/routes.ts:283]: Any reason for remove these from the children array?

File: apps/developer-portal/src/pages/user-stores-edit.tsx [hunk 18]
```diff
@@ -19,17 +19,20 @@
...
+    EditConnectionDetails,
+    EditGroupDetails,
+    EditUserDetails
 } from "../components";
+import { DatabaseAvatarGraphic } from "../configs";
 import { history } from "../helpers";
 import { PageLayout } from "../layouts"
-import { AlertLevels, Type, UserStore } from "../models";
+import { AlertLevels, CategorizedProperties, UserStore, UserstoreType } from "../models";
 import { addAlert } from "../store/actions";
```
Comment (brionmario) [inline apps/developer-portal/src/pages/user-stores-edit.tsx:34]:
```suggestion
import { addAlert } from "@wso2is/core/store";
```

**Pull request merged**